// @chiragrupani/karma-chromium-edge-launcher/dist/headlessChannels/EdgeBetaHeadless.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseBrowser_1 = __importDefault(require("../BaseBrowser"));
const Utilities_1 = __importDefault(require("../Utilities"));
const EdgeBetaHeadlessBrowser = function (baseBrowserDecorator, args) {
    baseBrowserDecorator(this);
    var flags = args.flags || [];
    var userDataDir = args.edgeDataDir || this._tempDir;
    var browser = new BaseBrowser_1.default(flags, userDataDir);
    this._getOptions = browser._getHeadlessOptions;
};
EdgeBetaHeadlessBrowser.$inject = ['baseBrowserDecorator', 'args'];
EdgeBetaHeadlessBrowser.prototype = {
    name: 'Edge Beta Headless',
    DEFAULT_CMD: {
        linux: Utilities_1.default.GetLinuxBin('microsoft-edge-beta'),
        darwin: Utilities_1.default.GetEdgeDarwin('/Applications/Microsoft Edge Beta.app/Contents/MacOS/Microsoft Edge Beta'),
        win32: Utilities_1.default.GetEdgeExe('Edge Beta'),
    },
    ENV_CMD: 'EDGE_BETA_BIN',
};
exports.default = EdgeBetaHeadlessBrowser;
