// @chiragrupani/karma-chromium-edge-launcher/dist/channels/EdgeCanary.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseBrowser_1 = __importDefault(require("../BaseBrowser"));
const Utilities_1 = __importDefault(require("../Utilities"));
const EdgeCanaryBrowser = function (baseBrowserDecorator, args) {
    baseBrowserDecorator(this);
    var flags = args.flags || [];
    var userDataDir = args.edgeDataDir || this._tempDir;
    var browser = new BaseBrowser_1.default(flags, userDataDir);
    this._getOptions = browser._getOptions;
};
EdgeCanaryBrowser.$inject = ['baseBrowserDecorator', 'args'];
EdgeCanaryBrowser.prototype = {
    name: 'Edge Canary',
    DEFAULT_CMD: {
        linux: Utilities_1.default.GetLinuxBin('microsoft-edge-canary'),
        darwin: Utilities_1.default.GetEdgeDarwin('/Applications/Microsoft Edge Canary.app/Contents/MacOS/Microsoft Edge Canary'),
        win32: Utilities_1.default.GetEdgeExe('Edge SxS'),
    },
    ENV_CMD: 'EDGE_CANARY_BIN',
};
exports.default = EdgeCanaryBrowser;
