// @pivotal-tools/commons-vscode/lib/code-lens-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const deepEqual = require("deep-equal");
const vscode_1 = require("vscode");
const highlight_service_1 = require("./highlight-service");
class HighlightCodeLensProvider {
    constructor() {
        this.highlights = new Map();
        this._onDidChangeCodeLenses = new vscode_1.EventEmitter();
    }
    get onDidChangeCodeLenses() {
        return this._onDidChangeCodeLenses.event;
    }
    handle(highlghtParams) {
        if (!deepEqual(this.highlights.get(highlghtParams.doc.uri), highlghtParams)) {
            this.highlights.set(highlghtParams.doc.uri, highlghtParams);
            this._onDidChangeCodeLenses.fire();
        }
    }
    static toVSCodeLens(cl) {
        const codeLens = {
            range: highlight_service_1.toVSRange(cl.range),
            isResolved: true,
            command: cl.command
        };
        return codeLens;
    }
    provideCodeLenses(document, token) {
        const activeUri = document.uri.toString();
        const activeVersion = document.version;
        const highlightParams = this.highlights.get(activeUri);
        if (highlightParams && highlightParams.doc.version === activeVersion) {
            const codeLenses = highlightParams.codeLenses || [];
            return codeLenses.filter(cl => cl.command).map(cl => HighlightCodeLensProvider.toVSCodeLens(cl));
        }
        return [];
    }
    ;
}
exports.HighlightCodeLensProvider = HighlightCodeLensProvider;
