// @chiragrupani/karma-chromium-edge-launcher/dist/channels/EdgeDev.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseBrowser_1 = __importDefault(require("../BaseBrowser"));
const Utilities_1 = __importDefault(require("../Utilities"));
const EdgeDevBrowser = function (baseBrowserDecorator, args) {
    baseBrowserDecorator(this);
    var flags = args.flags || [];
    var userDataDir = args.edgeDataDir || this._tempDir;
    var browser = new BaseBrowser_1.default(flags, userDataDir);
    this._getOptions = browser._getOptions;
};
EdgeDevBrowser.$inject = ['baseBrowserDecorator', 'args'];
EdgeDevBrowser.prototype = {
    name: 'Edge Dev',
    DEFAULT_CMD: {
        linux: Utilities_1.default.GetLinuxBin('microsoft-edge-dev'),
        darwin: Utilities_1.default.GetEdgeDarwin('/Applications/Microsoft Edge Dev.app/Contents/MacOS/Microsoft Edge Dev'),
        win32: Utilities_1.default.GetEdgeExe('Edge Dev'),
    },
    ENV_CMD: 'EDGE_DEV_BIN',
};
exports.default = EdgeDevBrowser;
