// svelte-language-server/dist/src/plugins/typescript/features/HoverProvider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HoverProviderImpl = void 0;
const typescript_1 = __importDefault(require("typescript"));
const documents_1 = require("../../../lib/documents");
const previewer_1 = require("../previewer");
const utils_1 = require("../utils");
const utils_2 = require("./utils");
class HoverProviderImpl {
    constructor(lsAndTsDocResolver) {
        this.lsAndTsDocResolver = lsAndTsDocResolver;
    }
    async doHover(document, position) {
        const { lang, tsDoc } = await this.getLSAndTSDoc(document);
        const fragment = await tsDoc.getFragment();
        const eventHoverInfo = await this.getEventHoverInfo(lang, document, tsDoc, fragment, position);
        if (eventHoverInfo) {
            return eventHoverInfo;
        }
        const offset = fragment.offsetAt(fragment.getGeneratedPosition(position));
        let info = lang.getQuickInfoAtPosition(tsDoc.filePath, offset);
        if (!info) {
            return null;
        }
        const textSpan = info.textSpan;
        // show docs of $store instead of store if necessary
        const is$store = fragment.text
            .substring(0, info.textSpan.start)
            .endsWith('(__sveltets_store_get(');
        if (is$store) {
            const infoFor$store = lang.getQuickInfoAtPosition(tsDoc.filePath, textSpan.start + textSpan.length + 3);
            if (infoFor$store) {
                info = infoFor$store;
            }
        }
        const declaration = typescript_1.default.displayPartsToString(info.displayParts);
        const documentation = previewer_1.getMarkdownDocumentation(info.documentation, info.tags);
        // https://microsoft.github.io/language-server-protocol/specification#textDocument_hover
        const contents = ['```typescript', declaration, '```']
            .concat(documentation ? ['---', documentation] : [])
            .join('\n');
        return documents_1.mapObjWithRangeToOriginal(fragment, {
            range: utils_1.convertRange(fragment, textSpan),
            contents
        });
    }
    async getEventHoverInfo(lang, doc, tsDoc, fragment, originalPosition) {
        const possibleEventName = documents_1.getWordAt(doc.getText(), doc.offsetAt(originalPosition), {
            left: /\S+$/,
            right: /[\s=]/
        });
        if (!possibleEventName.startsWith('on:')) {
            return null;
        }
        const component = await utils_2.getComponentAtPosition(this.lsAndTsDocResolver, lang, doc, tsDoc, fragment, originalPosition);
        if (!component) {
            return null;
        }
        const eventName = possibleEventName.substr('on:'.length);
        const event = component.getEvents().find((event) => event.name === eventName);
        if (!event) {
            return null;
        }
        return {
            contents: [
                '```typescript',
                `${event.name}: ${event.type}`,
                '```',
                event.doc || ''
            ].join('\n')
        };
    }
    async getLSAndTSDoc(document) {
        return this.lsAndTsDocResolver.getLSAndTSDoc(document);
    }
}
exports.HoverProviderImpl = HoverProviderImpl;
